sklearn/tests/test_discriminant_analysis.py

```diff
@@ -176,8 +175,12 @@ def test_lda_explained_variance_ratio():
     clf_lda_svd = LinearDiscriminantAnalysis(solver="svd")
     clf_lda_svd.fit(X, y)
     assert_almost_equal(clf_lda_svd.explained_variance_ratio_.sum(), 1.0, 3)
+
+    # NOTE: clf_lda_eigen.explained_variance_ratio_ is not of n_components
+    # length. Make it the same length as clf_lda_svd.explained_variance_ratio_
+    # before comparison.
     assert_array_almost_equal(clf_lda_svd.explained_variance_ratio_,
-                              clf_lda_eigen.explained_variance_ratio_)
+            clf_lda_eigen.explained_variance_ratio_[:3])
```

Comment:
I'd accept that length, if it were me.

By precision errors, I mean that the arrays you showed before

```
eigen
20
[  6.03795532e-01   3.96204468e-01   4.19383626e-16   3.08212423e-16
   1.23275796e-16   9.84308055e-17   6.96952083e-17   4.57224815e-17
   3.73609416e-17   2.12787319e-17   9.54458789e-18  -1.83915898e-17
  -4.11632284e-17  -1.08854065e-16  -1.41296422e-16  -1.58466942e-16
  -1.87047513e-16  -2.18274226e-16  -3.00368948e-16  -4.68795358e-16]
svd
3
[  6.03795532e-01   3.96204468e-01   1.23375797e-32]
Traceback (most recent call last):
  File "/home/djipey/informatique/python/scikit-learn/sklearn/tests/test_discriminant_analysis.py", line 368, in <module>
    test_lda_explained_variance_ratio()
  File "/home/djipey/informatique/python/scikit-learn/sklearn/tests/test_discriminant_analysis.py", line 193, in test_lda_explained_variance_ratio
    clf_lda_eigen.explained_variance_ratio_)
  File "/usr/lib/python3.5/site-packages/numpy/testing/utils.py", line 886, in assert_array_almost_equal
    precision=decimal)
  File "/usr/lib/python3.5/site-packages/numpy/testing/utils.py", line 663, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Arrays are not almost equal to 6 decimals
```

seem very solver dependent. The eigen solver might be more precise, and grabbing the eigenvalues that are extremely close to zero. 

Related topic, numerical stability:
https://en.wikipedia.org/wiki/Numerical_stability